### PR TITLE
Cleanup unecessary forwarding helpers in Expression

### PR DESCRIPTION
**Pull request by hughbe**

### CODE REVIEW

File: src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs [hunk 1]
```diff
@@ -171,7 +171,7 @@ public static Expression ValidateOneArgument(MethodBase method, ExpressionType n
             return arguments;
         }
 
-        public static void RequiresCanRead(Expression expression, string paramName, int idx)
+        public static void RequiresCanRead(Expression expression, string paramName, int idx = -1)
```
Comment (JonHanna) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Would it be worth doing this as an overload? On the one hand, the difference of having one extra argument in the calls is very small, but on the other hand this is called a great many times.

Comment (hughbe) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: I thought it was unnecessary, and dilutes the call stack especially considering exceptions are thrown here. Also, I actually prefer having the optional parameter - takes up less space, and makes the intention clear (that's just a style preference though)

Comment (VSadov) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Most of the callers seem to not pass `idx` and this helper is indeed called very often.
I think we should have an overload without `idx`.

**Comment by hughbe** [comment]
Friendly ping, let me know what you all think

**Comment by danmosemsft** [comment]
@VSadov @OmarTawfik to signoff.

**Comment by danmosemsft** [comment]
@VSadov @OmarTawfik @karelz

**Comment by karelz** [comment]
@VSadov @OmarTawfik ping again?

File: src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs [hunk 2]
```diff
@@ -171,6 +171,11 @@ public static Expression ValidateOneArgument(MethodBase method, ExpressionType n
             return arguments;
         }
 
+        public static void RequiresCanRead(Expression expression, string paramName)
+        {
+            RequiresCanRead(expression, paramName, -1);
+        }
```
Comment (VSadov) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Well, this is not much different from  passing the default "-1" and then checking for that same "-1" and ignoring it in downstream helpers.  
Is it possible to _not_ pass -1 and _not_ check for it  in helpers?

The point is that most common scenario is when the re is no index, so we should have a simpler codepath for those cases that does not involve the index at all.

Comment (hughbe) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Maybe. I think it'd take a lot of work and have little return. Taking a look at the code for RequiresCanRead, we pass the index to the exception throwing helpers: https://github.com/hughbe/corefx/blob/4f2fa6014e82ea87ffee20150cc0ce22d0a194ae/src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs#L190

This pattern looks unavoidable

Comment (JonHanna) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: I think this is a reasonable balance; it's worth having the overload for the slight size decrease it produces considering that it's called a great many times, but not worth having a full duplication within it.

Comment (hughbe) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Are you proposing keeping this as-is, right?

Comment (JonHanna) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: I am.

Comment (VSadov) [inline src/Common/src/System/Dynamic/Utils/ExpressionUtils.cs]: Ok, I guess at this point I have to agree.

**Pull request merged**